README.md
# Daedalus Poject's mysql-utils

The purpose of this is is avoid hardcodidng mysql commands inside Docker startup scripts.

[![pipeline status](https://git.daedalus-project.io/tools-and-utils/mysql-utils/badges/master/pipeline.svg)](https://git.daedalus-project.io/tools-and-utils/mysql-utils/commits/master)[![coverage report](https://git.daedalus-project.io/tools-and-utils/mysql-utils/badges/master/coverage.svg)](https://git.daedalus-project.io/tools-and-utils/mysql-utils/commits/master)

## Usage

```
Usage: daedalus-project-mysql-utils ACTION [ARGUMENTS]

When argumments are set and upercase like variables are set.
For example, setting --host="myhost" will set MYSQL_HOST to "myhost".

This script works with environment variables, if all required variables are set there is 
no need to set them using arguments.

Actions:
    usage:                                      Shows usage.
    change_root_password:                       Changes root's password.
    create_database:                            Creates databases.
    create_user:                                Creates users.
    grant:                                      Grants user access.

Arguments:

    Conection arguments:

    -u | --user=USER                            User wichs performs mysql connection.
                                                Sets MYSQL_USER

    -p | --password=PASSWORD                    User's password.
                                                Sets MYSQL_PASSWORD

    -h | --host=HOST                            Where msyql server is placed (localhost by default).
                                                Sets MYSQL_HOST

    -P | --post=PORT                            If this option is set this script uses host port for connections.
                                                Sets MYSQL_PORT  (default value is 3306)

    -S | --socket=SOCKET                        Socket to connect (/var/run/mysqld/mysqld.sock by default)
                                                Sets MYSQL_SOCKET

    -r | --connection-retries                   Number of times that this script tries to perfrom connection.
                                                Sets MYSQL_CONNECTION_RETRIES (default value is 5)

    -t | --connection-timeout                   Conection timeout in seconds
                                                Sets MYSQL_CONNECTION_TIMEOUT (default value is 10)


    change_root_password arguments:

    --new-root-password=PASS                    New root password
                                                Sets MYSQL_NEW_ROOT_PASSWORD

    --new-root-host                             root@'host'
                                                Sets MYSQL_NEW_ROOT_HOST

    create_database arguments:

    --database-name=NAME                        Name of the database to create
                                                Sets MYSQL_DATABASE_NAME

    create_user arguments:

    --new-user=NAME                             Name of the new user
                                                Sets MYSQL_NEW_USER

    --new-user-password=PASS                    Password of the new user
                                                Sets MYSQL_NEW_USER_PASSWORD

    --new-user-host=HOST                        New user's host
                                                Sets MYSQL_NEW_USER_HOST (% for example)

    grant arguments:

    --grant-priv-type=GRANT                     Grant priv type
                                                Sets MYSQL_GRANT_PRIV_TYPE (ALL PRIVILEGES by default)

    --grant-user=USER                           User to be granted
                                                Sets MYSQL_GRANT_USER

    --grant-database=DATABASE                   Database access to be granted
                                                Sets MYSQL_GRANT_DATABASE

    --grant-tables=TABLES                       Database tables to be granted
                                                Sets MYSQL_GRANT_TABLES (* by default)

    --grant-host=HOST                           User's host to be granted.
                                                Sets MYSQL_GRANT_HOST (% by default)

    --grant-other-account-characteristics=OTHER Extra characteristics
                                                Set MYSQL_GRANT_OTHER_ACCOUNT_CHARACTERISTICS
```


debian/changelog
daedalus-project-mysql-utils (0.4-7) RELEASED; urgency=medium

  * Update coverage

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Mon, 20 Feb 2020 19:45:00 +0100


daedalus-project-mysql-utils (0.4-6) RELEASED; urgency=medium

  * Use mysql_native_password

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Mon, 20 May 2019 18:31:00 +0200

daedalus-project-mysql-utils (0.4-5) RELEASED; urgency=medium

  * Main script was incomplete

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Mon, 13 May 2019 20:20:00 +0200

daedalus-project-mysql-utils (0.4-4) UNRELEASED; urgency=medium

  * Main script was incomplete

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Mon, 13 May 2019 20:20:00 +0200

daedalus-project-mysql-utils (0.4-3) UNRELEASED; urgency=medium

  * Main script was incomplete
  * New actions: usage

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Mon, 13 May 2019 07:37:00 +0200

daedalus-project-mysql-utils (0.4-2) UNRELEASED; urgency=medium

  * Main script was incomplete
  * New actions: create database, create user and grant

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Wed, 09 May 2019 18:17:00 +0200

daedalus-project-mysql-utils (0.4-1) UNRELEASED; urgency=medium

  * New actions: create database, create user and grant

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Wed, 08 May 2019 20:09:00 +0200

daedalus-project-mysql-utils (0.3-3) UNRELEASED; urgency=medium

  * New action: grant

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Wed, 07 May 2019 06:32:00 +0200

daedalus-project-mysql-utils (0.3-2) UNRELEASED; urgency=medium

  * New action: create user

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Wed, 06 May 2019 17:25:00 +0200

daedalus-project-mysql-utils (0.3-1) UNRELEASED; urgency=medium

  * New action: create database

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Wed, 05 May 2019 22:02:00 +0200

daedalus-project-mysql-utils (0.2-1) RELEASED; urgency=medium

  * First release
  * Available actions: change mysql root password

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Wed, 05 May 2019 17:36:00 +0200

daedalus-project-mysql-utils (0.1-2) UNRELEASED; urgency=medium

  * Manage LOCAL_ERROR_FILE

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Wed, 05 May 2019 16:45:00 +0200

daedalus-project-mysql-utils (0.1-1) UNRELEASED; urgency=medium

  * First release
  * Available actions: change mysql root password

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Wed, 05 May 2019 13:47:00 +0200


lib/00-variables.sh
#!/bin/bash -
#===============================================================================
#
#          FILE: 00-variables.sh
#
#   DESCRIPTION: Global Variables

#        AUTHOR: Álvaro Castellano Vela (alvaro.castellano.vela@gmail.com)
#  ORGANIZATION: Daedalus Project
#       CREATED: 08/05/2019 18:09
#===============================================================================

declare -A connection_options_hash=(
["u"]="user"
["p"]="password"
["h"]="host"
["P"]="port"
["S"]="socket"
["r"]="connection-retries"
["t"]="connection-timeout"
)

declare -a change_root_password_options_array=(
"new-root-password"
"new-root-host"
)

declare -a create_database_options_array=(
"database-name"
)

declare -a create_user_options_array=(
"new-user"
"new-user-password"
"new-user-host"
)

declare -a grant_options_array=(
"grant-priv-type"
"grant-user"
"grant-database"
"grant-host"
"grant-tables"
"grant-other-account-characteristics"
)

declare -A default_connection_options_hash=(
["MYSQL_HOST"]="localhost"
["MYSQL_SOCKET"]="/var/run/mysqld/mysqld.sock"
["MYSQL_CONNECTION_RETRIES"]=5
["MYSQL_CONNECTION_TIMEOUT"]=10 # seconds
)

declare required_variables=(
"MYSQL_USER"
"MYSQL_PASSWORD"
"MYSQL_HOST"
#"MYSQL_PORT" - Not required
"MYSQL_SOCKET"
"MYSQL_CONNECTION_RETRIES"
"MYSQL_CONNECTION_TIMEOUT"
)

declare -A default_grant_options_hash=(
["MYSQL_GRANT_PRIV_TYPE"]="ALL PRIVILEGES"
["MYSQL_GRANT_OTHER_ACCOUNT_CHARACTERISTICS"]=""
["MYSQL_GRANT_HOST"]="%"
["MYSQL_GRANT_TABLES"]="*"
)

declare -a available_actions=(
'change_root_password'
'create_database'
'create_user'
'grant'
'usage'
)
